research/minigo/minigo.py

```diff
@@ -0,0 +1,406 @@
+    params: Hyperparameters.
+
+  Returns:
+    The model name of the winner.
+  """
+
+  black_model = os.path.join(models_dir, black_model_name)
+  white_model = os.path.join(models_dir, white_model_name)
+
+  print('Evaluate models between %s and %s' % (
+      black_model_name, white_model_name))
+
+  _ensure_dir_exists(evaluate_dir)
+
+  with utils.logged_timer('Loading weights'):
+    black_net = dual_net.DualNetwork(black_model, params)
+    white_net = dual_net.DualNetwork(white_model, params)
+
+  with utils.logged_timer('%d games' % params.eval_games):
+    winner = evaluation.play_match(
+        black_net, white_net, params.eval_games, params.eval_readouts,
+        evaluate_dir, params.eval_verbose)
+
+  return black_model_name if winner == 'B' else white_model_name
+
+
+def main(_):
+  """Run the reinforcement learning loop."""
+  tf.logging.set_verbosity(tf.logging.INFO)
+
+  params = model_utils.MiniGoModelParams()
+
+  k = utils.round_power_of_two(FLAGS.go_size ** 2 / 3)
+  params.num_filters = k  # Number of filters in the convolution layer
+  params.fc_width = 2 * k  # Width of each fully connected layer
+  params.num_shared_layers = FLAGS.go_size  # Number of shared trunk layers
+  params.go_size = FLAGS.go_size  # Board size
+
+  # set the shuffle buffer size smaller
+  preprocessing.SHUFFLE_BUFFER_SIZE = 200000
+
+  # How many positions can fit on a graphics card. 256 for 9s, 16 or 32 for 19s.
+  if FLAGS.go_size == 9:
+    params.batch_size = 256
+  else:
+    params.batch_size = 32
+
+  # A dummy model for debug/testing purpose with fewer games and iterations
+  if FLAGS.debug:
+    params.num_filters = 8  # Number of filters in the convolution layer
+    params.fc_width = 16  # Width of each fully connected layer
+    params.num_shared_layers = 1  # Number of shared trunk layers
+    params.batch_size = 16
+    params.examples_per_generation = 64
+    params.max_games_per_generation = 10
+    params.max_iters_per_pipeline = 2
+    preprocessing.SHUFFLE_BUFFER_SIZE = 1000
+
+  # Set directories for models and datasets
+  if os.path.isdir(FLAGS.base_dir):  # if it already exists, delete it.
+    shutil.rmtree(FLAGS.base_dir, ignore_errors=True)
+  models_dir = os.path.join(FLAGS.base_dir, 'models')
+  estimator_working_dir = os.path.join(FLAGS.base_dir, 'estimator_working_dir/')
+  selfplay_dir = os.path.join(FLAGS.base_dir, 'data/selfplay/')
+  holdout_dir = os.path.join(FLAGS.base_dir, 'data/holdout/')
+  training_chunk_dir = os.path.join(FLAGS.base_dir, 'data/training_chunks/')
+  sgf_dir = os.path.join(FLAGS.base_dir, 'sgf/')
+  evaluate_dir = os.path.join(FLAGS.base_dir, 'sgf/evaluate/')
+
+  print('Creating random initial weights...')
+  bootstrap(estimator_working_dir, models_dir, params)
+
+  # Start from the bootstrap model to play games.
+  print('Playing some games...')
+  model_name = '000000-bootstrap'
+  for rl_iter in range(params.max_iters_per_pipeline):
```

Comment:
We should have options to only do the selfplay, or only do the train-eval... this will allow us to achieve process level paralellization easily.